### PR TITLE
[UX] Filter choices for field order when creating Custom Fields

### LINKED ISSUE #14180: "Field order" should display only other custom fields when creating a Custom Field
### Mautic Series

5.1.x series

### Mautic installed version

5.x branch

### Way of installing

I cloned the Mautic repo (not recommended or supported)

### PHP version

Not relevant

### What browsers are you seeing the problem on?

Chrome

### What happened?

Every time you create a new custom field, you have the option to select where to place it (you'd see it later when editing a Contact or Company). But the system doesn't take into account that the Core fields that come by default with Mautic cannot be sorted. 

This means that, if you select, for example, the option "Company name" to place your new custom field above this one, it will simply not happen.

The new custom field will be placed at the bottom of the editing page for the corresponding object, after all Core fields are rendered.

So, if these default fields are fixed, they must not appear as an option for the "Field order". When was select existing custom fields, you can successfully place the new field above the selected field, but that's the only possible situation. 

To fix, "Field order" must display only other existing custom fields, hiding the system default (Core) fields.

### How can we reproduce this issue?

1. Open Custom Fields
2. Create New
3. Click on the "Field order" field and see that Core (system defaults) fields are displayed and cause no changes when used

### Relevant log output

_No response_

### Code of Conduct

- [X] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

### PR DESCRIPTION
**Pull request by laurielim**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | 🟢 
| New feature/enhancement? (use the a.x branch)      | 🟢 - enhancement
| Deprecations?                          | 🔴
| BC breaks? (use the c.x branch)        | 🔴
| Automated tests included?              | 🔴 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #14180 & #14181 

## Description
This PR solves issues #14180 & #14181 with some additional related changes
- Updated order field query_builder so that choices are filtered by selected object and group
- Added Ajax so that the choice list is updated when object or group changes, which involved refactoring how the field is added
- Disabled Field order of Fixed field and set tooltip accordingly
- Order the fields in quick add form so that Fixed fields come first

Note the reorderFieldsByEntity method might be buggy, and sometimes the fields are placed below instead of above selected field. Not sure what is causing it and might need some guidance if it needs to be fixed in this PR.

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to custom fields and add a couple of custom fields
3. See that the options in field order changes based on the object and group selected
4. Once you've added  custom fields, check the form by going to contacts/companies and clicking on new
5. Edit a newly created field and see that you can change the group and the order
6. Go to existing fields that cannot be reordered and see that the field order is disabled and the tooltip states that default field cannot be reordered
7. Create a new field where "Visible on quick add" is checked and see that the field is placed below the fixed fields such as first name and last name

### CODE REVIEW

**Review by andersonjeccel — APPROVED** [review]
Works perfectly

Companies:
image

Contacts:
image

Thanks for your contribution <3

**Comment by andersonjeccel** [comment]
@laurielim  Oh, we just need some final steps:

- Rector flagged something in your code, adjusting to the suggestion will fix it _(small issue, one line)_
- It lowered a bit of the coverage and the maintainers will ask for a PHPUnit test, is that something you could include?

Ping me if you have questions

**Comment by andersonjeccel** [comment]
@escopecz Looking at the tests, they seem all OK
Codecov is also passing

Is this ready to be merged?

**Comment by laurielim** [comment]
Fixed the rector issue. About tests, I don't really have experience with unit tests so if they're needed, I could give it a go if I could get some guidance / examples I could look at.

**Comment by andersonjeccel** [comment]
@laurielim 
These are docs: https://docs.phpunit.de/en/8.5/
This is the current test file: app/bundles/LeadBundle/Tests/Controller/AjaxControllerFunctionalTest.php

From what I understood, you'll have to update the test to cover only the lines

Other reference that might help (OpenAI's o1-mini model): https://chatgpt.com/share/672e1e2a-0398-8007-8338-f670575f964c

**Review by shinde-rahul — CHANGES_REQUESTED** [review]
Code changes look good to me. Please add the coverage to complete the PR.

**Comment by andersonjeccel** [comment]
@abhisekmazumdar Is there a possibility for you to help this contributor to complete the PHPUnit test?

My skills cover only CSS, HTML and Twig, so I’m not able to help :/

**Comment by abhisekmazumdar** [comment]
I will take a look and try to make it possible to move.

**Comment by laurielim** [comment]
I pushed a commit containing a functional test for the ajax action if you can maybe review that?

**Comment by abhisekmazumdar** [comment]
@laurielim I can't recreate the failed test cases locally. Can you update the branch and see if the tests still fail?

**Comment by abhisekmazumdar** [comment]
Checking the e2e.

**Comment by laurielim** [comment]
@abhisekmazumdar the e2e is known to fail from what i understand. I think what @andersonjeccel meant when tagging you is if you can check the test i added and see if it's good and if there's more to be done.

**Comment by abhisekmazumdar** [comment]
Aah! I got lost. :-(

I tested the new test case added in @app/bundles/LeadBundle/Tests/Controller/AjaxControllerFunctionalTest.php@.

With my limited understanding of the tests, I feel it covers the requirements well. However, I will leave it to others to determine if there are any additional corner cases that need to be addressed.

**Comment by andersonjeccel** [comment]
Everything seems fine now

I’m so happy we got this ready to be merged

A huge THANK YOU again @laurielim 🙏🏻

Also thanks @abhisekmazumdar for having a look!

File: app/bundles/LeadBundle/Tests/Controller/AjaxControllerFunctionalTest.php
Comment (escopecz) [inline app/bundles/LeadBundle/Tests/Controller/AjaxControllerFunctionalTest.php]: This could be simplified with PHPUNIT's data providers. What do you think?
```suggestion
    /**
     * @dataProvider leadFieldOrderChoiceListProvider
     *
     * @param string[] $expectedOptions
     */
    public function testUpdateLeadFieldOrderChoiceListAction(string $object, string $group, array $expectedOptions): void
    {
        $payload = [
            'action' => 'lead:updateLeadFieldOrderChoiceList',
            'object' => $object,
            'group'  => $group,
        ];

        $this->client->request(
            Request::METHOD_POST,
            '/s/ajax',
            $payload,
            [],
            $this->createAjaxHeaders()
        );

        // Get the response HTML
        $response    = $this->client->getResponse();
        $htmlContent = $response->getContent();

        // Assert the response is successful
        $this->assertTrue($response->isOk(), "Response was not OK for object: $object, group: $group");
        $this->assertStringNotContainsString('<form', $htmlContent, 'Response contains a form instead of just field order.');
        $this->assertStringContainsString('<select', $htmlContent, 'Response contains select tag.');

        // Parse the HTML content using DOMDocument
        $dom = new \DOMDocument();
        @$dom->loadHTML($htmlContent);
        $select  = $dom->getElementsByTagName('select')->item(0);
        $options = $select->getElementsByTagName('option');

        $actualOptions = [];
        foreach ($options as $option) {
            if ($option->textContent) {
                // Get the text content of each <option>
                $actualOptions[] = trim($option->textContent);
            }
        }
        // Assert that the actual options match the expected options
        if (empty($expectedOptions)) {
            $this->assertEmpty($actualOptions);
        }
        foreach ($expectedOptions as $expectedValue) {
            $this->assertContains($expectedValue, $actualOptions, "Missing expected option '$expectedValue' for object: $object, group: $group");
        }
    }

    public function leadFieldOrderChoiceListProvider(): \Generator
    {
        yield ['lead', 'core', ['Fax', 'Website']];
        yield ['lead', 'social', ['Facebook', 'Foursquare', 'Instagram']];
        yield ['company', 'core', []];
    }
```

Comment (laurielim) [reply app/bundles/LeadBundle/Tests/Controller/AjaxControllerFunctionalTest.php]: Yes it definitely simplifies it, didn't realise I could do that, thanks!

**Review by escopecz — APPROVED** [review]
Thanks! The code looks good 👍

**Pull request merged**
